File: AIQuotaBar/Services/Clash/ClashRouteFilter.swift
import Foundation

struct ClashRouteFilterResult: Equatable, Sendable {
    let routes: [ClashRoute]
    let errorMessage: String?
}

enum ClashRouteFilter {
    static func filter(
        _ routes: [ClashRoute],
        query: String,
        usesRegularExpression: Bool
    ) -> ClashRouteFilterResult {
        let trimmedQuery = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedQuery.isEmpty else {
            return ClashRouteFilterResult(routes: routes, errorMessage: nil)
        }

        if usesRegularExpression {
            do {
                let expression = try NSRegularExpression(
                    pattern: trimmedQuery,
                    options: [.caseInsensitive])
                let filtered = routes.filter { route in
                    let range = NSRange(route.name.startIndex..., in: route.name)
                    return expression.firstMatch(
                        in: route.name,
                        options: [],
                        range: range) != nil
                }
                return ClashRouteFilterResult(routes: filtered, errorMessage: nil)
            } catch {
                return ClashRouteFilterResult(
                    routes: [],
                    errorMessage: error.localizedDescription)
            }
        }

        let normalizedQuery = normalize(trimmedQuery)
        let queryCountries = Set(countryAliases.compactMap { country in
            country.aliases.contains(where: { normalize($0) == normalizedQuery })
                ? country.code
                : nil
        })

        let filtered = routes.filter { route in
            let normalizedName = normalize(route.name)
            if normalizedName.contains(normalizedQuery) {
                return true
            }

            guard !queryCountries.isEmpty else { return false }
            let routeCountries = Set(countryAliases.compactMap { country in
                country.aliases.contains(where: { alias in
                    containsAlias(alias, in: route.name)
                }) ? country.code : nil
            })
            return !routeCountries.isDisjoint(with: queryCountries)
        }

        return ClashRouteFilterResult(routes: filtered, errorMessage: nil)
    }

    private static func normalize(_ value: String) -> String {
        value
            .folding(
                options: [.caseInsensitive, .diacriticInsensitive, .widthInsensitive],
                locale: Locale(identifier: "en_US_POSIX"))
            .lowercased()
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func containsAlias(_ alias: String, in name: String) -> Bool {
        let normalizedAlias = normalize(alias)
        let normalizedName = normalize(name)

        guard normalizedAlias.range(of: #"^[a-z]{2,3}$"#, options: .regularExpression) != nil else {
            return normalizedName.contains(normalizedAlias)
        }

        guard let expression = try? NSRegularExpression(
            pattern: "(^|[^a-z])\(NSRegularExpression.escapedPattern(for: normalizedAlias))([^a-z]|$)",
            options: [.caseInsensitive]) else {
            return false
        }
        let range = NSRange(normalizedName.startIndex..., in: normalizedName)
        return expression.firstMatch(in: normalizedName, range: range) != nil
    }

    private static let countryAliases: [CountryAliases] = [
        CountryAliases(code: "JP", aliases: ["🇯🇵", "日本", "Japan", "JP"]),
        CountryAliases(code: "SG", aliases: ["🇸🇬", "新加坡", "Singapore", "SG"]),
        CountryAliases(code: "US", aliases: ["🇺🇸", "美国", "United States", "USA", "US"]),
        CountryAliases(code: "HK", aliases: ["🇭🇰", "香港", "Hong Kong", "HK"]),
        CountryAliases(code: "TW", aliases: ["🇹🇼", "台湾", "臺灣", "Taiwan", "TW"]),
        CountryAliases(code: "KR", aliases: ["🇰🇷", "韩国", "韓國", "Korea", "KR"]),
        CountryAliases(code: "GB", aliases: ["🇬🇧", "英国", "英國", "United Kingdom", "UK", "GB"]),
        CountryAliases(code: "DE", aliases: ["🇩🇪", "德国", "德國", "Germany", "DE"]),
        CountryAliases(code: "FR", aliases: ["🇫🇷", "法国", "法國", "France", "FR"]),
        CountryAliases(code: "CA", aliases: ["🇨🇦", "加拿大", "Canada", "CA"]),
        CountryAliases(code: "AU", aliases: ["🇦🇺", "澳洲", "澳大利亚", "Australia", "AU"]),
        CountryAliases(code: "NL", aliases: ["🇳🇱", "荷兰", "荷蘭", "Netherlands", "NL"]),
    ]
}

private struct CountryAliases {
    let code: String
    let aliases: [String]
}

File: windows/src/AIQuotaBar.Providers/Clash/ClashRouteFilter.cs
// Swift 来源：AIQuotaBar/Services/Clash/ClashRouteFilter.swift — enum ClashRouteFilter + private CountryAliases（v1.28.1）
// 对应测试：AIQuotaBar/Tests/Clash/ClashRouteFilterTests.swift

#nullable enable

using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;

namespace AIQuotaBar.Providers.Clash;

/// <summary>
/// 线路搜索过滤器：普通模式按归一化子串 + 国家别名（旗帜/中文/ISO 码）匹配；
/// 正则模式直接匹配原始线路名（忽略大小写），非法模式返回错误信息与空结果。
/// </summary>
public static class ClashRouteFilter
{
    private static readonly Regex ShortCodeExpression = new("^[a-z]{2,3}$", RegexOptions.Compiled);

    public static ClashRouteFilterResult Filter(
        IReadOnlyList<ClashRoute> routes,
        string query,
        bool usesRegularExpression)
    {
        var trimmedQuery = query.Trim();
        if (trimmedQuery.Length == 0)
        {
            return new ClashRouteFilterResult(routes, null);
        }

        if (usesRegularExpression)
        {
            try
            {
                var expression = new Regex(trimmedQuery, RegexOptions.IgnoreCase);
                var filtered = routes
                    .Where(route => expression.IsMatch(route.Name))
                    .ToList();
                return new ClashRouteFilterResult(filtered, null);
            }
            catch (RegexParseException exception)
            {
                return new ClashRouteFilterResult(
                    Array.Empty<ClashRoute>(),
                    exception.Message);
            }
        }

        var normalizedQuery = Normalize(trimmedQuery);
        var queryCountries = CountryAliasTable
            .Where(country => country.Aliases.Any(alias => Normalize(alias) == normalizedQuery))
            .Select(country => country.Code)
            .ToHashSet();

        var filteredRoutes = routes
            .Where(route =>
            {
                if (Normalize(route.Name).Contains(normalizedQuery, StringComparison.Ordinal))
                {
                    return true;
                }

                if (queryCountries.Count == 0)
                {
                    return false;
                }

                var routeCountries = CountryAliasTable
                    .Where(country => country.Aliases.Any(alias => ContainsAlias(alias, route.Name)))
                    .Select(country => country.Code)
                    .ToHashSet();
                return routeCountries.Overlaps(queryCountries);
            })
            .ToList();

        return new ClashRouteFilterResult(filteredRoutes, null);
    }

    /// <summary>
    /// Swift folding(.caseInsensitive + .diacriticInsensitive + .widthInsensitive) 的 .NET 镜像：
    /// FormKC 折叠全半角，FormD 后剥离组合记号再去音符号，最后小写 + 去首尾空白。
    /// </summary>
    private static string Normalize(string value)
    {
        var widthFolded = value.Normalize(NormalizationForm.FormKC);
        var decomposed = widthFolded.Normalize(NormalizationForm.FormD);
        var builder = new StringBuilder(decomposed.Length);
        foreach (var character in decomposed)
        {
            if (CharUnicodeInfo.GetUnicodeCategory(character) != UnicodeCategory.NonSpacingMark)
            {
                builder.Append(character);
            }
        }

        return builder
            .ToString()
            .Normalize(NormalizationForm.FormC)
            .ToLowerInvariant()
            .Trim();
    }

    private static bool ContainsAlias(string alias, string name)
    {
        var normalizedAlias = Normalize(alias);
        var normalizedName = Normalize(name);

        // 短别名（2-3 个字母的 ISO 码）要求词边界，避免 "SG" 命中 "USGS-01"；
        // 其余别名（旗帜、中文等）按归一化子串匹配。
        if (!ShortCodeExpression.IsMatch(normalizedAlias))
        {
            return normalizedName.Contains(normalizedAlias, StringComparison.Ordinal);
        }

        var pattern = $"(^|[^a-z]){Regex.Escape(normalizedAlias)}([^a-z]|$)";
        var expression = new Regex(pattern, RegexOptions.IgnoreCase);
        return expression.IsMatch(normalizedName);
    }

    private static readonly CountryAliases[] CountryAliasTable =
    {
        new("JP", new[] { "🇯🇵", "日本", "Japan", "JP" }),
        new("SG", new[] { "🇸🇬", "新加坡", "Singapore", "SG" }),
        new("US", new[] { "🇺🇸", "美国", "United States", "USA", "US" }),
        new("HK", new[] { "🇭🇰", "香港", "Hong Kong", "HK" }),
        new("TW", new[] { "🇹🇼", "台湾", "臺灣", "Taiwan", "TW" }),
        new("KR", new[] { "🇰🇷", "韩国", "韓國", "Korea", "KR" }),
        new("GB", new[] { "🇬🇧", "英国", "英國", "United Kingdom", "UK", "GB" }),
        new("DE", new[] { "🇩🇪", "德国", "德國", "Germany", "DE" }),
        new("FR", new[] { "🇫🇷", "法国", "法國", "France", "FR" }),
        new("CA", new[] { "🇨🇦", "加拿大", "Canada", "CA" }),
        new("AU", new[] { "🇦🇺", "澳洲", "澳大利亚", "Australia", "AU" }),
        new("NL", new[] { "🇳🇱", "荷兰", "荷蘭", "Netherlands", "NL" }),
    };

    private sealed record CountryAliases(string Code, IReadOnlyList<string> Aliases);
}
